### PR TITLE
perf: only fetch limited messages (pagination)

### LINKED ISSUE #446: Remove Virtuoso from Raven Web app
After a careful regression test, I figured out that Virtuoso has been the main reason why the chat interface lags so much.

1. From the Virtuoso Docs: https://virtuoso.dev/#gotchas - adding CSS margins makes the virtualization process slower
2. https://virtuoso.dev/#performance - Performance is impacted by dynamic heights, widths and a large viewport. In our case, our message items span across 80-90% of the screen + all our content is dynamic (markdown, images, files). There's no way for any virtualization engine to figure out the heights of these items without actually mounting them.

We need to revert Virtuoso and bring back our simple, yet humble, chat interface.

### LINKED ISSUE #462: [RAV-18] Socket events should mutate the message API if the sender is from a different device
Current logic:
If the sender of the message is self, do not mutate since we already mutated when we sent the message.

Drawback:
User sends the message from mobile, but the message is not updated on the web app since the "sender" is the same. We need to also pass in the Session ID to check if we should mutate or not. 

<sub>[RAV-18](https://linear.app/the-commit-company/issue/RAV-18/socket-events-should-mutate-the-message-api-if-the-sender-is-from-a)</sub>

### LINKED ISSUE #567: [RAV-3] Add skeleton loader for messages on web app
null

<sub>[RAV-3](https://linear.app/the-commit-company/issue/RAV-3/add-skeleton-loader-for-messages-on-web-app)</sub>

### LINKED ISSUE #625: User avatar is showing skeleton on scroll
https://github.com/The-Commit-Company/Raven/assets/141011354/0d88d955-2a12-4ff7-95b1-c62b23f645d1

### PR DESCRIPTION
**Pull request by nikkothari22**
api: previous and new message fetching API
perf: send message reactions in websocket
perf: send deleted message in websocket
perf: send edited message in websocket
perf: send saved message event in websocket
perf: send new messages in websocket (closes #462)
fix: User avatar is showing skeleton on scroll (closes #625)
feat: highlight replied message after scroll
perf: render action palette with hover debounce

Changes only made on the web app so far. Mobile app and desk integration needs to be changed.

### Pending items:

- [x] Scroll to message (by search, saved messages or reply) which has not been loaded in the initial page
- [x] API to fetch newer messages needs to be tested
- [x] Track channel visit if latest messages are fetched
- [x] Track channel visit if the user is focused and new message comes in via websockets
- [x] Do not show unread count if user is on the channel
- [x] Date separators needs to be added back. Format dates in it to avoid expensive renders
- [x] Remove Virtuoso library and it's files (closes #446 )
- [x] Show button when new messages are available
- [x] Add skeleton loader for messages (closes #567)

Will keep adding to this thread as things progress.

### CODE REVIEW

**Comment by nikkothari22** [comment]
Closes #625

**Comment by nikkothari22** [comment]
image
 Skeleton loader for web app messages

**Comment by nikkothari22** [comment]
Even better solution: No mutations, just rely on websockets and SWR for everything (yeah). 

Instead of maintaining two arrays for messages and previous messages, we can maintain a single array
This is maintained by the useSWR hook (useFrappeGetCall).

We use the mutate method to update the messages array when messages are received, updated, deleted, etc.
Even if the user scrolls up and loads older messages, the hook is mutated by concatenating the older messages to the existing messages array

Since we are using Web Socket events, this saves multiple round-trips to the server to fetch messages (pre Raven v1.5)

When the page changes and the user comes back to the chat, the messages are fetched again and hence the messages array is updated with only new messages

This also helps when the user goes directly to a specific message from somewhere(notification, search etc.). 
We just shift the base message and fire the `get_messages` API to get the messages around the base message.
Post that, the behaviour is the same - if the user scrolls up or down, we just mutate the messages array directly.
If a new messages comes in while the user is viewing the base message (i.e. the latest messages are not visible), we do not update the messages array.
This is tracked using the `has_new_messages` flag in the response from the `get_messages` API

Below are the Web Socket events and they are all handled by mutating the messages array directly

Refer: https://swr.vercel.app/docs/mutation

File: raven-app/src/components/feature/chat/ChatStream/ChatStream.tsx [hunk 1]
```diff
@@ -0,0 +1,115 @@
+import { Message } from '../../../../../../types/Messaging/Message'
+import { DeleteMessageDialog, useDeleteMessage } from '../ChatMessage/MessageActions/DeleteMessage'
+import { EditMessageDialog, useEditMessage } from '../ChatMessage/MessageActions/EditMessage'
+import { MessageItem } from '../ChatMessage/MessageItem'
+import { ChannelHistoryFirstMessage } from '@/components/layout/EmptyState'
+import { useParams } from 'react-router-dom'
+import useChatStream from './useChatStream'
+import { UIEventHandler, useEffect, useRef, useState } from 'react'
+import { Flex, ScrollArea } from '@radix-ui/themes'
```
Comment (TITANiumRox) [inline raven-app/src/components/feature/chat/ChatStream/ChatStream.tsx:9]:
```suggestion
import { Flex } from '@radix-ui/themes'
```

**Review by TITANiumRox — APPROVED** [review]

**Comment by nikkothari22** [comment]
Issue with file uploads here: two websocket events are published - one without a file, the other with.

**Comment by nikkothari22** [comment]
#### Unread message counts:

For every channel member, we store a `last_visit` timestamp in the Raven Channel Member doctype. To get the number of unread messages, we can simply look at the no. of messages created after this timestamp for any given channel.

The `last_visit` for a member of a channel is updated when:
1. Latest messages are fetched (usually when a channel is opened)- this is in the `get_messages` API under `chat_stream`
2. The user fetches newer messages (let's say they were viewing a saved message (older) and scrolled down until they reached the bottom of the chat stream). This is in the `get_newer_messages` API under `chat_stream`.
3. When the user sends a new message (handled by controller under Raven Message)
4. When the user exits a channel after all the latest messages had been loaded

When Raven loads, we fetch the unread message counts for all channels. Post that, updates to these counts are made when:
1. If a user opens a channel - we locally update the unread message count to 0 - no API call
2. If a realtime event is published for unread message count change and the sender is not the user itself - we only fetch the unread count for the particular channel (instead of all channels like we used to).

The realtime event for unread message count changed is published when:
1. A new message is sent
2. A message is deleted
3. The user scrolls to the bottom of the chat stream from a base message.

**Comment by nikkothari22** [comment]
Closes #771

**Pull request merged**